synapse/storage/controllers/state.py

```diff
@@ -182,6 +182,56 @@ def _get_state_groups_from_groups(
 
         return self.stores.state._get_state_groups_from_groups(groups, state_filter)
 
+    @trace
+    @tag_args
+    async def _get_state_for_events_when_filtering_for_client(
+        self, event_ids: Collection[str], user_id_viewing_events: str
+    ) -> Dict[str, StateMap[EventBase]]:
+        """Get the state at each event that is necessary to filter
+        them before being displayed to clients from the perspective of the
+        `user_id_viewing_events`. Will fetch `m.room.history_visibility` and
+        `m.room.member` event of `user_id_viewing_events`.
+
+        Args:
+            event_ids: List of event ID's that will be displayed to the client
+            user_id_viewing_events: User ID that will be viewing these events
+
+        Returns:
+            Dict of event_id to state map.
+        """
+        set_tag(
+            SynapseTags.FUNC_ARG_PREFIX + "event_ids.length",
+            str(len(event_ids)),
+        )
+
+        # Since we're making decisions based on the state, we need to wait.
+        await_full_state = True
```

Comment:
Is this accurate?

Reply:
Probably the source of our `TestPartialStateJoin` failures in Complement (https://github.com/matrix-org/synapse/actions/runs/3520947916/jobs/5902365185#step:6:11950). What should I be doing here?